src/libraries/System.Private.Uri/src/System/PercentEncodingHelper.cs

```diff
@@ -0,0 +1,158 @@
+﻿using System.Buffers;
+using System.Diagnostics;
+using System.Text;
+
+namespace System
+{
+    internal static class PercentEncodingHelper
+    {
+        public static unsafe int UnescapePercentEncodedUTF8Sequence(char* input, int length, ref ValueStringBuilder dest, bool isQuery, bool iriParsing)
+        {
+            // As an optimization, this method should only be called after the first character is known to be a part of a non-ascii UTF8 sequence
+            Debug.Assert(length >= 3);
+            Debug.Assert(input[0] == '%');
+            Debug.Assert(UriHelper.EscapedAscii(input[1], input[2]) != Uri.c_DummyChar);
+            Debug.Assert(UriHelper.EscapedAscii(input[1], input[2]) >= 128);
+
+            uint fourByteBuffer = 0;
+            int bytesLeftInBuffer = 0;
+            var fourByteSpan = new ReadOnlySpan<byte>(&fourByteBuffer, 4);
+
+            int totalCharsConsumed = 0;
+            int charsToCopy = 0;
+            int charsConsumed = 0;
+
+        RefillBuffer:
+            int i = totalCharsConsumed + (bytesLeftInBuffer * 3);
+
+        ReadByteFromInput:
+            if ((uint)(i + 2) >= length || input[i] != '%')
+                goto NoMoreOrInvalidInput;
+
+            uint value = input[i + 1];
+            if ((uint)((value - 'A') & ~0x20) <= ('F' - 'A'))
+            {
+                value = (value | 0x20) - 'a' + 10;
+            }
+            else if ((value - '8') <= ('9' - '8'))
```

Comment:
You could use your helpers in the other file for this.

Reply:
I manually inlined it here, as we are only interested in non-ascii ones